src/orc.jl

```diff
@@ -0,0 +1,151 @@
+export OrcJIT, OrcModule, OrcTargetAddress
+export JITEventListener, GDBRegistrationListener, IntelJITEventListener,
+       OProfileJITEventListener, PerfJITEventListener
+export dispose, errormsg, compile!, remove!, add!,
+       mangle, address, addressin, create_stub!, set_stub!,
+       register!, unregister!, callback!
+export JITTargetMachine
+
+@checked struct OrcJIT 
+    ref::API.LLVMOrcJITStackRef
+end
+
+Base.unsafe_convert(::Type{API.LLVMOrcJITStackRef}, orc::OrcJIT) = orc.ref
+
+"""
+    OrcJIT(::TargetMachine)
+
+Creates a OrcJIT stack based on the provided target machine.
+
+!!! warning
+    Takes ownership of the provided target machine.
+"""
+function OrcJIT(tm::TargetMachine)
+    OrcJIT(API.LLVMOrcCreateInstance(tm))
+end
+
+function dispose(orc::OrcJIT)
+    API.LLVMOrcDisposeInstance(orc)
+end
+
+function errormsg(orc::OrcJIT)
+    unsafe_string(LLVM.API.LLVMOrcGetErrorMsg(orc))
+end
+
+struct OrcModule 
+    handle::API.LLVMOrcModuleHandle
+end
+Base.cconvert(::Type{API.LLVMOrcModuleHandle}, mod::OrcModule) = mod
+Base.unsafe_convert(::Type{API.LLVMOrcModuleHandle}, mod::OrcModule) = mod.handle
+
+function compile!(orc::OrcJIT, mod::Module, resolver = C_NULL, ctx = C_NULL; lazy=false)
+    r_mod = Ref{API.LLVMOrcModuleHandle}()
+    if lazy
+        API.LLVMOrcAddLazilyCompiledIR(orc, r_mod, mod, resolver, ctx)
+    else
+        API.LLVMOrcAddEagerlyCompiledIR(orc, r_mod, mod, resolver, ctx)
+    end
+    OrcModule(r_mod[])
+end
+
+function remove!(orc::OrcJIT, mod::OrcModule)
+    LLVM.API.LLVMOrcRemoveModule(orc, mod)
+end
+
+function add!(orc::OrcJIT, obj::MemoryBuffer, resolver = C_NULL, ctx = C_NULL)
+    r_mod = Ref{API.LLVMOrcModuleHandle}()
+    API.LLVMOrcAddObjectFile(orc, r_mod, obj, resolver, ctx)
+    return OrcModule(r_mod[])
+end
+
+function mangle(orc::OrcJIT, name)
+    r_symbol = Ref{Cstring}()
+    LLVM.API.LLVMOrcGetMangledSymbol(orc, r_symbol, name)
+    symbol = unsafe_string(r_symbol[])
+    LLVM.API.LLVMOrcDisposeMangledSymbol(r_symbol[])
+    return symbol
+end
+
+struct OrcTargetAddress
+    ptr::API.LLVMOrcTargetAddress
+end
+Base.cconvert(::Type{API.LLVMOrcTargetAddress}, addr::OrcTargetAddress) = addr
```

Comment:
Why not just override `Base.convert`? We're not really taking the address of anything here, so it isn't an unsafe conversion. Then you can also drop the `cconvert`.